Testing FedAvgGrad

danielspg opened the pull request:
Created a Strategy named FedAvgGrad:

the FedAvgGrad follows the following sample structure:
![FedAvgGrad](https://user-images.githubusercontent.com/19836648/126508711-9b694c99-2da2-43d9-b948-4fb7991e50e1.jpg)

Created a example which utilizes the FedAvgGrad strategy:
A test run is shown below:
![FedAvgGradTest](https://user-images.githubusercontent.com/19836648/126509650-f035d0db-6ad4-4c7b-bab9-51523c2c6367.png)

pedropgusmao reviewed (commented):
Hello @danielspg  Thanks for your contribution. Would mind include the tests for this strategy as well? You can use https://github.com/adap/flower/blob/main/src/py/flwr/server/strategy/fedavg_test.py for reference .

danielspg commented:
1. Modified [FedAvgGrad](https://github.com/danielspg/flower/blob/testing_FedAvgGrad/src/py/flwr/server/strategy/FedAvgGrad.py) strategy code to accommodate DEPRECATIONS of Strategy.aggregate_fit(...) and some other functions in Flower.

2. Fixed the accuracy issue in example: [pytorch_mnist_FedAvgGrad](https://github.com/danielspg/flower/tree/testing_FedAvgGrad/examples/pytorch_mnist_FedAvgGrad)
Below is the execution of example: Using 2 clients and training for 5 round. Final accuracy was around 97%
![MNIST_FedAvgGrad](https://user-images.githubusercontent.com/19836648/131459494-a874bf9c-5c4f-40d1-8153-5066d22f6dab.png)


3. **Added Example utilizing FedAvgGrad strategy based on CIFAR dataset: [quickstart_pytorch_FedAvgGrad](https://github.com/danielspg/flower/tree/testing_FedAvgGrad/examples/quickstart_pytorch_FedAvgGrad)**
Below is the execution of example: Using 2 clients and training for 3 rounds. Final Accuracy was around 43%
![Cifar_FedAvgGrad](https://user-images.githubusercontent.com/19836648/131460850-c814e42b-b383-4637-82b4-2966c586bcc8.png)

jafermarq commented:
Hi @danielspg, sorry for the long delay with your PR. We are still happy to move ahead with your contribution but we were thinking to start with just the `FedAvgGrad` strategy. This could be the building block for subsequent strategies that operate with gradients (understood as the difference between the global model received by clients and the locally update model) as opposed to models.

Are you still interested in integrating `FedAvgGrad` into Flower?
